### PR TITLE
fix MNIST byte flipping

### LINKED ISSUE #7079: The labels in the QMNIST dataset are all equal to 0.
### 🐛 Describe the bug

The labels in the QMNIST dataset are all equal to 0.
```python
im_transform = T.Compose([
    T.ToTensor()
])

qmnist_train = torchvision.datasets.QMNIST('data/', train=True, download=True, transform=im_transform)

# the first image is a '5'
plt.imshow(qmnist_train[0][0][0])
# prints '0'
print(qmnist_train[0][1])

# the following never outputs anything
[print(qmnist_train[i][1]) for i in range(len(qmnist_train)) if qmnist_train[i][1] != 0]
```

### Versions

```
Collecting environment information...
PyTorch version: 1.13.1
Is debug build: False
CUDA used to build PyTorch: 11.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.1 LTS (x86_64)
GCC version: (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.35

Python version: 3.10.8 (main, Nov 24 2022, 14:13:03) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-57-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 11.6.124
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce GTX 980 Ti
Nvidia driver version: 525.60.11
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] geotorch==0.3.0
[pip3] numpy==1.23.5
[pip3] torch==1.13.1
[pip3] torchaudio==0.13.1
[pip3] torchvision==0.14.1
[conda] blas                      1.0                         mkl  
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] geotorch                  0.3.0                    pypi_0    pypi
[conda] mkl                       2021.4.0           h06a4308_640  
[conda] mkl-service               2.4.0           py310h7f8727e_0  
[conda] mkl_fft                   1.3.1           py310hd6ae3a3_0  
[conda] mkl_random                1.2.2           py310h00e6091_0  
[conda] numpy                     1.23.5          py310hd5efca6_0  
[conda] numpy-base                1.23.5          py310h8e6c178_0  
[conda] pytorch                   1.13.1          py3.10_cuda11.6_cudnn8.3.2_0    pytorch
[conda] pytorch-cuda              11.6                 h867d48c_1    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torchaudio                0.13.1              py310_cu116    pytorch
[conda] torchvision               0.14.1              py310_cu116    pytorch
```

cc @pmeier

### PR DESCRIPTION
**Pull request by pmeier**
Fixes #7079.

### CODE REVIEW

**Review by NicolasHug — COMMENTED** [review]
Thanks Philip.
Before merging, could we add some quick test for the endianness conversion (possibly requiring extracting it in a separate function)?

```py
t = torch.tensor([0x00112233, 0xaabbccdd], dtype=float.torch32)
flipped = torch.tensor([0x33221100, 0xddccbbaa], dtype=float.torch32)
assert flip_endianness(t) == flipped
```

Ideally we'd test the QMnist dataset directly, but this may make the tests just as complicated as the code (which kinds defeats the purpose of the test)

File: test/test_datasets_utils.py [hunk 1]
```diff
@@ -215,6 +217,26 @@ def test_verify_str_arg(self):
         pytest.raises(ValueError, utils.verify_str_arg, 0, ("a",), "arg")
         pytest.raises(ValueError, utils.verify_str_arg, "b", ("a",), "arg")
 
+    @pytest.mark.parametrize(
+        ("dtype", "actual_hex", "expected_hex"),
+        [
+            (torch.uint8, "01 23 45 67 89 AB CD EF", "01 23 45 67 89 AB CD EF"),
+            (torch.float16, "01 23 45 67 89 AB CD EF", "23 01 67 45 AB 89 EF CD"),
+            (torch.int32, "01 23 45 67 89 AB CD EF", "67 45 23 01 EF CD AB 89"),
+            (torch.float64, "01 23 45 67 89 AB CD EF", "EF CD AB 89 67 45 23 01"),
```
Comment (NicolasHug) [inline test/test_datasets_utils.py:226]: Not critical, but I'd prefer creating the tensors objects directly here; one shouldn't need to understand `frombuffer` nor `fromhex` to understand this test.

**Review by NicolasHug — APPROVED** [review]
Thanks @pmeier , some comments but LGTM

File: test/test_datasets_utils.py (hunk 1 above)
Comment (pmeier) [reply test/test_datasets_utils.py:226]: As discussed offline, the direct way, e.g. `torch.tensor([0x00112233], dtype=torch.float32)` is not possible for floating point inputs. Although written as hexadecimal, the value is an integer and so `torch` converts it to a `float` which garbles the bytes. The only way to actually preserve the bytes regardless of the dtype is by going through `torch.frombuffer`.
